lib/extreme/persistent_subscription.ex

```diff
@@ -0,0 +1,325 @@
+    end
+
+    def init(opts) do
+      {:ok, opts}
+    end
+
+    def subscribe(listener_proc), do: GenServer.cast(listener_proc, :subscribe)
+
+    def handle_cast(:subscribe, state) do
+      {:ok, subscription_pid} =
+        MyExtremeClientModule.connect_to_persistent_subscription(
+          self(),
+          opts.stream,
+          opts.group,
+          opts.allowed_in_flight_messages
+        )
+
+      {:noreply, Map.put(state, :subscription_pid, subscription_pid)}
+    end
+
+    def handle_cast({:on_event, event, correlation_id}, state) do
+      # .. do the real processing here ..
+
+      :ok = PersistentSubscription.ack(state.subscription_pid, event, correlation_id)
+
+      {:noreply, state}
+    end
+
+    def handle_cast(:unsubscribe, state) do
+      :ok = PersistentSubscription.unsubscribe(state.subscription_pid)
+
+      {:noreply, state}
+    end
+
+    def handle_info(_, state), do: {:noreply, state}
+  end
+  ```
+  """
+
+  use GenServer
+  require Logger
+  alias Extreme.SharedSubscription, as: Shared
+  alias Extreme.RequestManager
+  alias Extreme.Messages, as: Msg
+
+  defmodule State do
+    @moduledoc false
+
+    defstruct ~w(
+      base_name
+      correlation_id
+      subscriber
+      stream
+      group
+      allowed_in_flight_messages
+      status
+      subscription_id
+    )a
+  end
+
+  @typedoc """
+  An event received from a persistent subscription.
+  """
+  @type event() :: %Extreme.Messages.ResolvedIndexedEvent{}
+  @typedoc """
+  An event ID.
+
+  Either from the `:link` or `:event`, depending on if the event is from a
+  projection stream or a normal stream (respectively).
+  """
+  @type event_id() :: binary()
+
+  @doc false
+  def start_link(
+        base_name,
+        correlation_id,
+        subscriber,
+        stream,
+        group,
+        allowed_in_flight_messages
+      ) do
+    GenServer.start_link(
+      __MODULE__,
+      {base_name, correlation_id, subscriber, stream, group, allowed_in_flight_messages}
+    )
+  end
+
+  @doc """
+  Acknowledges that an event or set of events have been successfully processed.
+
+  `ack/3` takes any of the following for event ID:
+
+  * a full event, as given in the `{:on_event, event, correlation_id}` cast
+  * the `event_id` of an event (either from its `:link` or `:event`, depending
+    on if the event comes from a projection or a normal stream, respectively)
+  * a list of either sort
+
+  `correlation_id` comes from the `:on_event` cast.
+
+  ## Example
+
+      def handle_cast({:on_event, event, correlation_id}, state) do
+        # .. do some processing ..
+
+        # when the processing completes successfully:
+        :ok = Extreme.PersistentSubscription.ack(state.subscription_pid, event, correlation_id)
+
+        {:noreply, state}
+      end
+  """
+  @spec ack(pid(), event() | event_id() | [event() | event_id()], binary()) :: :ok
+  def ack(subscription, event, correlation_id) when is_map(event) or is_binary(event) do
```

Comment:
I think we could eliminate `correlation_id` as a parameter here because `correlation_id` will always be `state.correlation_id` wrt the subscription process

so really it's redundant to provide it to `ack/3` and `nack/5`

unless a listener subscribes to more than one persistent subscription. Because the events are delivered with `GenServer.cast/2`, the `from` is not available, the correlation ID is the only way to distinguish which subscription an event come from.

I don't think it's a good idea to have one listener subscribe to multiple persistent subscriptions, but I don't wanna force anyone to not have that option

`¯\_(ツ)_/¯` leaving off that refactor for now, but open to debate

Reply:
I agree that one listener shouldn't subscribe to multiple persistent subscriptions and I would actually discourage that.